File: src/player_search.py
"""Player search algorithm for finding DUPR ratings."""

import re
import sys
from dataclasses import dataclass
from typing import Optional, List, Tuple

from .config import Config, debug_log
from .dupr_client import DUPRClient, DUPRPlayer, DUPRAPIError, TokenExpiredError


@dataclass
class SearchResult:
    """Result of a player search."""
    name: str
    rating: float
    player_id: Optional[int]
    profile_url: Optional[str]
    found: bool
    search_method: str  # Describes how the player was found


# Common short last names that need special handling (full name search preferred)
SHORT_COMMON_LASTNAMES = {
    'ng', 'hu', 'wu', 'li', 'le', 'lu', 'ma', 'xu', 'yu', 'ye', 'he', 'ho',
    'wong', 'chen', 'wang', 'zhang', 'liu', 'yang', 'huang', 'zhao', 'zhou', 'sun'
}


class PlayerSearcher:
    """Searches for players using the defined algorithm."""

    def __init__(self, config: Config, client: DUPRClient):
        self.config = config
        self.client = client

    def _clean_name(self, name: str) -> str:
        """
        Clean player name by removing guest markers and other annotations.

        Removes:
        - (G) or (g) - guest marker
        - (Guest) - full guest marker
        - Trailing parenthetical annotations
        """
        # Remove common guest markers
        cleaned = re.sub(r'\s*\([Gg](uest)?\)\s*', '', name)
        # Remove any other trailing parenthetical content
        cleaned = re.sub(r'\s*\([^)]*\)\s*$', '', cleaned)
        return cleaned.strip()

    def _normalize_name(self, name: str) -> str:
        """Normalize a name for comparison."""
        return name.lower().strip()

    def _first_name_matches(self, search_first: str, api_first: str) -> bool:
        """Check if the API first name contains the search first name."""
        search_normalized = self._normalize_name(search_first)
        api_normalized = self._normalize_name(api_first)
        return search_normalized in api_normalized or api_normalized in search_normalized

    def _find_unique_match(
        self,
        players: List[DUPRPlayer],
        first_name: str,
        full_name: str
    ) -> Optional[DUPRPlayer]:
        """Find a unique match in the player list.

        Matching priority:
        1. If only one result, return it
        2. Look for exact full name match (case-insensitive)
        3. Fall back to first name matching if no exact match
        """
        if len(players) == 1:
            return players[0]

        # First: Look for exact full name match
        full_name_normalized = self._normalize_name(full_name)
        exact_matches = [
            p for p in players
            if self._normalize_name(p.full_name) == full_name_normalized
        ]

        if len(exact_matches) == 1:
            debug_log(f"Found exact full name match: {exact_matches[0].full_name}")
            return exact_matches[0]

        if len(exact_matches) > 1:
            debug_log(f"Found {len(exact_matches)} exact full name matches for '{full_name}'")
            # Multiple exact matches - can't determine which one, return first
            return exact_matches[0]

        # Fallback: First name matching (for cases where full name doesn't exactly match)
        matches = [
            p for p in players
            if self._first_name_matches(first_name, p.first_name)
        ]

        if len(matches) == 1:
            debug_log(f"Found unique first name match: {matches[0].full_name}")
            return matches[0]

        debug_log(f"No unique match found: {len(matches)} first name matches for '{first_name}'")
        return None

    def _search_with_filter(
        self,
        query: str,
        first_name: str,
        full_name: str,
        location_text: Optional[str],
        lat: Optional[float],
        lng: Optional[float],
        filter_desc: str
    ) -> Tuple[Optional[DUPRPlayer], str]:
        """Perform a search with the given filter settings."""
        debug_log(f"Searching '{query}' with filter: {filter_desc}")

        try:
            players = self.client.search_players(
                query=query,
                location_text=location_text,
                lat=lat,
                lng=lng
            )
        except DUPRAPIError as e:
            debug_log(f"API error during search: {e}")
            return None, ""

        match = self._find_unique_match(players, first_name, full_name)
        if match:
            return match, f"{filter_desc}"

        return None, ""

    def _is_short_common_lastname(self, name: str) -> bool:
        """Check if a name is a short common last name that needs special handling."""
        return name.lower() in SHORT_COMMON_LASTNAMES

    def search_player(self, full_name: str) -> SearchResult:
        """
        Search for a player using the defined algorithm.

        Search sequence:
        1. Check player_overrides.json (using original and cleaned names)
        2. Full Name + Alberta filter
        3. Last Name + Alberta filter (skip for very common short last names)
        4. Full Name + Canada filter
        5. Last Name + Canada filter
        6. Last Name + No filter
        7. Fallback to default rating
        """
        # Check override with original name first
        name_key = self._normalize_name(full_name)
        if name_key in self.config.overrides:
            override = self.config.overrides[name_key]
            debug_log(f"Using override for '{full_name}': {override.rating} ({override.reason})")
            return SearchResult(
                name=full_name,
                rating=override.rating,
                player_id=None,
                profile_url=None,
                found=True,
                search_method=f"Override: {override.reason}"
            )

        # Clean the name (remove guest markers, etc.)
        cleaned_name = self._clean_name(full_name)
        if cleaned_name != full_name:
            debug_log(f"Cleaned name: '{full_name}' -> '{cleaned_name}'")

        # Check override with cleaned name if different
        cleaned_key = self._normalize_name(cleaned_name)
        if cleaned_key != name_key and cleaned_key in self.config.overrides:
            override = self.config.overrides[cleaned_key]
            debug_log(f"Using override for cleaned name '{cleaned_name}': {override.rating} ({override.reason})")
            return SearchResult(
                name=full_name,
                rating=override.rating,
                player_id=None,
                profile_url=None,
                found=True,
                search_method=f"Override: {override.reason}"
            )

        # Parse cleaned name
        name_parts = cleaned_name.strip().split()
        first_name = name_parts[0] if name_parts else ""
        last_name = name_parts[-1] if len(name_parts) > 1 else name_parts[0] if name_parts else ""

        # Determine if this is a short common last name
        is_short_lastname = self._is_short_common_lastname(last_name)
        if is_short_lastname:
            debug_log(f"Short common last name detected: '{last_name}'")

        try:
            # Step 2: Full Name + Alberta filter (primary search)
            match, method = self._search_with_filter(
                query=cleaned_name,
                first_name=first_name,
                full_name=cleaned_name,
                location_text=self.config.ALBERTA_TEXT,
                lat=self.config.ALBERTA_LAT,
                lng=self.config.ALBERTA_LNG,
                filter_desc="Full name + Alberta"
            )
            if match:
                return self._create_result(full_name, match, method)

            # Step 3: Last Name + Alberta filter
            # Skip for very common short last names as they return too many results
            if not is_short_lastname:
                match, method = self._search_with_filter(
                    query=last_name,
                    first_name=first_name,
                    full_name=cleaned_name,
                    location_text=self.config.ALBERTA_TEXT,
                    lat=self.config.ALBERTA_LAT,
                    lng=self.config.ALBERTA_LNG,
                    filter_desc="Last name + Alberta"
                )
                if match:
                    return self._create_result(full_name, match, method)

            # Step 4: Full Name + Canada filter (for players outside Alberta)
            match, method = self._search_with_filter(
                query=cleaned_name,
                first_name=first_name,
                full_name=cleaned_name,
                location_text=self.config.CANADA_TEXT,
                lat=self.config.CANADA_LAT,
                lng=self.config.CANADA_LNG,
                filter_desc="Full name + Canada"
            )
            if match:
                return self._create_result(full_name, match, method)

            # Step 5: Last Name + Canada filter
            if not is_short_lastname:
                match, method = self._search_with_filter(
                    query=last_name,
                    first_name=first_name,
                    full_name=cleaned_name,
                    location_text=self.config.CANADA_TEXT,
                    lat=self.config.CANADA_LAT,
                    lng=self.config.CANADA_LNG,
                    filter_desc="Last name + Canada"
                )
                if match:
                    return self._create_result(full_name, match, method)

            # Step 6: Last Name + No filter (global search)
            if not is_short_lastname:
                match, method = self._search_with_filter(
                    query=last_name,
                    first_name=first_name,
                    full_name=cleaned_name,
                    location_text=None,
                    lat=None,
                    lng=None,
                    filter_desc="Last name + No filter"
                )
                if match:
                    return self._create_result(full_name, match, method)

            # Step 6b: Full Name + No filter (last resort for short last names)
            match, method = self._search_with_filter(
                query=cleaned_name,
                first_name=first_name,
                full_name=cleaned_name,
                location_text=None,
                lat=None,
                lng=None,
                filter_desc="Full name + No filter"
            )
            if match:
                return self._create_result(full_name, match, method)

        except TokenExpiredError:
            # Re-raise token errors to halt execution
            raise

        except DUPRAPIError as e:
            debug_log(f"API error searching for '{full_name}': {e}")

        # Step 7: Fallback
        print(f"Warning: Player '{full_name}' not found, using default rating", file=sys.stderr)
        return SearchResult(
            name=full_name,
            rating=self.config.DEFAULT_RATING,
            player_id=None,
            profile_url=None,
            found=False,
            search_method="Default (player not found)"
        )

    def _create_result(self, search_name: str, player: DUPRPlayer, method: str) -> SearchResult:
        """Create a SearchResult from a matched player."""
        rating = player.best_rating
        if rating is None:
            rating = self.config.DEFAULT_RATING
            debug_log(f"Player '{player.full_name}' has no rating, using default")

        return SearchResult(
            name=search_name,
            rating=rating,
            player_id=player.id,
            profile_url=player.profile_url,
            found=True,
            search_method=method
        )

File: packages/core/src/player-search.ts
/**
 * Player Search Module
 * Implements 8-tier cascade search for finding DUPR players
 */

import { DUPRClient, DUPRPlayer, TokenExpiredError } from './dupr-client';
import { PlayerRegistry } from './player-registry';
import { NicknameResolver, nicknameResolver } from './nickname-resolver';

// =============================================================================
// Constants
// =============================================================================

export const ALBERTA_LAT = 53.9332706;
export const ALBERTA_LNG = -116.5765035;
export const CANADA_LAT = 56.130366;
export const CANADA_LNG = -106.346771;
export const DEFAULT_RATING = 3.0;
export const FUZZY_THRESHOLD = 0.85;

// Common short last names that need special handling (full name search preferred)
const SHORT_COMMON_LASTNAMES = new Set([
  'ng', 'hu', 'wu', 'li', 'le', 'lu', 'ma', 'xu', 'yu', 'ye', 'he', 'ho',
  'wong', 'chen', 'wang', 'zhang', 'liu', 'yang', 'huang', 'zhao', 'zhou', 'sun',
  'smith', 'lee', 'kim'
]);

// =============================================================================
// Types
// =============================================================================

export interface SearchResult {
  /** Original search name */
  name: string;
  /** Player's rating (DEFAULT_RATING if not found) */
  rating: number;
  /** DUPR player ID if found */
  playerId: number | null;
  /** DUPR ID string if found */
  duprId: string | null;
  /** Full name from DUPR if found */
  duprName: string | null;
  /** Profile URL if found */
  profileUrl: string | null;
  /** Whether player was found in DUPR */
  found: boolean;
  /** How the player was found (registry, override, exact_match, fuzzy_match, default) */
  searchMethod: string;
  /** Location from DUPR if found */
  location: string | null;
}

export interface PlayerOverride {
  duprId: string;
  duprName: string;
  rating: number;
}

export interface SearchConfig {
  /** Player overrides by lowercase name */
  overrides?: Map<string, PlayerOverride>;
  /** Location text for search (e.g., "Alberta, Canada") */
  defaultLocationText?: string;
}

/** Progress callback for batch search operations */
export interface SearchProgressCallback {
  /** Called when starting to search for a player */
  onStart?: (index: number, total: number, name: string) => void;
  /** Called when a player search completes */
  onComplete?: (index: number, total: number, result: SearchResult) => void;
}

// =============================================================================
// PlayerSearcher Class
// =============================================================================

export class PlayerSearcher {
  private client: DUPRClient;
  private registry: PlayerRegistry;
  private nicknameResolver: NicknameResolver;
  private overrides: Map<string, PlayerOverride>;
  private defaultLocationText: string;
  private debug: boolean;

  constructor(
    client: DUPRClient,
    registry?: PlayerRegistry,
    config?: SearchConfig
  ) {
    this.client = client;
    this.registry = registry ?? new PlayerRegistry();
    this.nicknameResolver = nicknameResolver;
    this.overrides = config?.overrides ?? new Map();
    this.defaultLocationText = config?.defaultLocationText ?? 'Alberta, Canada';
    this.debug = true;
  }

  private log(...args: unknown[]): void {
    if (this.debug) {
      console.log('[PlayerSearcher]', ...args);
    }
  }

  /**
   * Search for a player using 8-tier cascade
   *
   * Cascade order:
   * 1. Registry (cached matches)
   * 2. Overrides
   * 3. Full Name + Alberta location
   * 4. Last Name + Alberta location
   * 5. Full Name + Canada location
   * 6. Last Name + Canada location
   * 7. Last Name + No filter
   * 8. Full Name + No filter
   * 9. Default rating fallback
   */
  async searchPlayer(fullName: string): Promise<SearchResult> {
    const cleanName = this.cleanName(fullName);
    const nameLower = cleanName.toLowerCase();

    this.log(`Searching for player: "${cleanName}"`);

    // Tier 1: Check registry cache
    const cached = this.registry.get(nameLower);
    if (cached) {
      this.log(`  Found in registry cache: ${cached.duprName} (${cached.rating})`);
      return this.makeResult(cleanName, cached.rating ?? DEFAULT_RATING, {
        playerId: parseInt(cached.duprId) || null,
        duprId: cached.duprId,
        duprName: cached.duprName,
        profileUrl: `https://dashboard.dupr.com/dashboard/player/${cached.duprId}`,
        found: true,
        searchMethod: 'registry',
        location: cached.location,
      });
    }

    // Tier 2: Check overrides
    const override = this.overrides.get(nameLower);
    if (override) {
      this.log(`  Found in overrides: ${override.duprName} (${override.rating})`);
      return this.makeResult(cleanName, override.rating, {
        playerId: null,
        duprId: override.duprId,
        duprName: override.duprName,
        profileUrl: null,
        found: true,
        searchMethod: 'override',
        location: null,
      });
    }

    // Extract name parts
    const parts = cleanName.split(/\s+/);
    const lastName = parts[parts.length - 1];
    const isShortCommonLastName = this.isShortCommonName(lastName);

    // Tier 3: Full Name + Alberta
    this.log(`  Tier 3: Searching "${cleanName}" in Alberta...`);
    let result = await this.searchWithLocation(cleanName, 'Alberta, Canada', ALBERTA_LAT, ALBERTA_LNG);
    if (result) {
      this.log(`  Found via exact_alberta: ${result.fullName} (${result.bestRating})`);
      this.cacheResult(cleanName, result);
      return this.makeFoundResult(cleanName, result, 'exact_alberta');
    }

    // Tier 4: Last Name + Alberta (skip for short common names)
    if (!isShortCommonLastName && lastName !== cleanName) {
      this.log(`  Tier 4: Searching last name "${lastName}" in Alberta...`);
      result = await this.searchWithLocation(lastName, 'Alberta, Canada', ALBERTA_LAT, ALBERTA_LNG, cleanName);
      if (result) {
        this.log(`  Found via lastname_alberta: ${result.fullName} (${result.bestRating})`);
        this.cacheResult(cleanName, result);
        return this.makeFoundResult(cleanName, result, 'lastname_alberta');
      }
    }

    // Tier 5: Full Name + Canada
    this.log(`  Tier 5: Searching "${cleanName}" in Canada...`);
    result = await this.searchWithLocation(cleanName, 'Canada', CANADA_LAT, CANADA_LNG);
    if (result) {
      this.log(`  Found via exact_canada: ${result.fullName} (${result.bestRating})`);
      this.cacheResult(cleanName, result);
      return this.makeFoundResult(cleanName, result, 'exact_canada');
    }

    // Tier 6: Last Name + Canada
    if (!isShortCommonLastName && lastName !== cleanName) {
      this.log(`  Tier 6: Searching last name "${lastName}" in Canada...`);
      result = await this.searchWithLocation(lastName, 'Canada', CANADA_LAT, CANADA_LNG, cleanName);
      if (result) {
        this.log(`  Found via lastname_canada: ${result.fullName} (${result.bestRating})`);
        this.cacheResult(cleanName, result);
        return this.makeFoundResult(cleanName, result, 'lastname_canada');
      }
    }

    // Tier 7: Last Name + No filter
    if (!isShortCommonLastName && lastName !== cleanName) {
      this.log(`  Tier 7: Searching last name "${lastName}" globally...`);
      result = await this.searchNoFilter(lastName, cleanName);
      if (result) {
        this.log(`  Found via lastname_global: ${result.fullName} (${result.bestRating})`);
        this.cacheResult(cleanName, result);
        return this.makeFoundResult(cleanName, result, 'lastname_global');
      }
    }

    // Tier 8: Full Name + No filter
    this.log(`  Tier 8: Searching "${cleanName}" globally...`);
    result = await this.searchNoFilter(cleanName);
    if (result) {
      this.log(`  Found via exact_global: ${result.fullName} (${result.bestRating})`);
      this.cacheResult(cleanName, result);
      return this.makeFoundResult(cleanName, result, 'exact_global');
    }

    // Tier 9: Default rating fallback
    this.log(`  NOT FOUND - using default rating ${DEFAULT_RATING}`);
    return this.makeResult(cleanName, DEFAULT_RATING, {
      playerId: null,
      duprId: null,
      duprName: null,
      profileUrl: null,
      found: false,
      searchMethod: 'default',
      location: null,
    });
  }

  /**
   * Search with location filter and find best match
   */
  private async searchWithLocation(
    query: string,
    locationText: string,
    lat: number,
    lng: number,
    fullName?: string
  ): Promise<DUPRPlayer | null> {
    try {
      const players = await this.client.searchPlayers(query, locationText, lat, lng);
      this.log(`    API returned ${players.length} results for "${query}" in ${locationText}`);
      return this.findBestMatch(players, fullName ?? query);
    } catch (error) {
      // Re-throw auth errors so UI can handle them
      if (error instanceof TokenExpiredError) {
        throw error;
      }
      const message = error instanceof Error ? error.message : 'Unknown error';
      this.log(`    ERROR searching "${query}" in ${locationText}: ${message}`);
      return null;
    }
  }

  /**
   * Search without location filter
   */
  private async searchNoFilter(query: string, fullName?: string): Promise<DUPRPlayer | null> {
    try {
      const players = await this.client.searchPlayers(query);
      this.log(`    API returned ${players.length} results for "${query}" (no location)`);
      return this.findBestMatch(players, fullName ?? query);
    } catch (error) {
      // Re-throw auth errors so UI can handle them
      if (error instanceof TokenExpiredError) {
        throw error;
      }
      const message = error instanceof Error ? error.message : 'Unknown error';
      this.log(`    ERROR searching "${query}" globally: ${message}`);
      return null;
    }
  }

  /**
   * Find the best matching player from search results
   * Uses exact matching, then nickname matching, then fuzzy matching
   */
  private findBestMatch(players: DUPRPlayer[], searchName: string): DUPRPlayer | null {
    if (players.length === 0) return null;

    const searchLower = searchName.toLowerCase();
    const searchParts = searchLower.split(/\s+/);

    // Try exact match first
    for (const player of players) {
      if (player.fullName.toLowerCase() === searchLower) {
        return player;
      }
    }

    // Try nickname matching
    for (const player of players) {
      const playerParts = player.fullName.toLowerCase().split(/\s+/);
      if (this.namesMatchWithNicknames(searchParts, playerParts)) {
        return player;
      }
    }

    // Try fuzzy matching
    for (const player of players) {
      if (this.nicknameResolver.fuzzyMatch(searchName, player.fullName, FUZZY_THRESHOLD)) {
        return player;
      }
    }

    // Return first result if any exist (best API match)
    return players[0];
  }

  /**
   * Check if names match considering nicknames
   */
  private namesMatchWithNicknames(searchParts: string[], playerParts: string[]): boolean {
    if (searchParts.length !== playerParts.length) return false;

    for (let i = 0; i < searchParts.length; i++) {
      const search = searchParts[i];
      const player = playerParts[i];

      if (search === player) continue;
      if (this.nicknameResolver.areNamesEquivalent(search, player)) continue;
      return false;
    }

    return true;
  }

  /**
   * Clean player name (remove guest markers, trim whitespace)
   */
  private cleanName(name: string): string {
    return name
      .replace(/\s*\(G\)\s*/gi, '')
      .replace(/\s*\(Guest\)\s*/gi, '')
      .replace(/\s*\([^)]*\)\s*$/, '')
      .trim();
  }

  /**
   * Check if a name is short and common (skip last-name-only searches)
   */
  private isShortCommonName(name: string): boolean {
    return name.length <= 3 || SHORT_COMMON_LASTNAMES.has(name.toLowerCase());
  }

  /**
   * Cache a successful search result
   */
  private cacheResult(searchName: string, player: DUPRPlayer): void {
    this.registry.register(
      searchName,
      player.duprId,
      player.fullName,
      player.bestRating,
      player.shortAddress
    );
  }

  /**
   * Create a search result from a found player
   */
  private makeFoundResult(name: string, player: DUPRPlayer, method: string): SearchResult {
    return this.makeResult(name, player.bestRating ?? DEFAULT_RATING, {
      playerId: player.id,
      duprId: player.duprId,
      duprName: player.fullName,
      profileUrl: player.profileUrl,
      found: true,
      searchMethod: method,
      location: player.shortAddress,
    });
  }

  /**
   * Create a search result
   */
  private makeResult(
    name: string,
    rating: number,
    data: Omit<SearchResult, 'name' | 'rating'>
  ): SearchResult {
    return { name, rating, ...data };
  }

  /**
   * Add or update an override
   */
  setOverride(name: string, override: PlayerOverride): void {
    this.overrides.set(name.toLowerCase(), override);
  }

  /**
   * Remove an override
   */
  removeOverride(name: string): boolean {
    return this.overrides.delete(name.toLowerCase());
  }

  /**
   * Get the registry (for persistence)
   */
  getRegistry(): PlayerRegistry {
    return this.registry;
  }

  /**
   * Get the DUPR client (for testing/access)
   */
  getClient(): DUPRClient {
    return this.client;
  }

  /**
   * Search for multiple players
   * @param names Array of player names to search
   * @param progress Optional callback for progress updates
   * @returns Array of search results in same order as input
   */
  async searchPlayers(names: string[], progress?: SearchProgressCallback): Promise<SearchResult[]> {
    this.log(`Starting search for ${names.length} players`);
    const results: SearchResult[] = [];
    let foundCount = 0;
    let notFoundCount = 0;

    for (let i = 0; i < names.length; i++) {
      const name = names[i];
      this.log(`[${i + 1}/${names.length}] Processing: "${name}"`);

      // Notify progress callback that we're starting this player
      progress?.onStart?.(i, names.length, name);

      const result = await this.searchPlayer(name);
      results.push(result);

      if (result.found) {
        foundCount++;
      } else {
        notFoundCount++;
      }

      // Notify progress callback that this player is complete
      progress?.onComplete?.(i, names.length, result);
    }

    this.log(`Search complete: ${foundCount} found, ${notFoundCount} not found (using default ${DEFAULT_RATING})`);
    return results;
  }
}
